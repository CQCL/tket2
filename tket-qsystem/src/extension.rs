//! This module defines the Hugr extensions used by tket-qsystem.
pub mod futures;
pub mod qsystem;
pub mod random;
pub mod result;
pub mod utils;
pub mod wasm;
