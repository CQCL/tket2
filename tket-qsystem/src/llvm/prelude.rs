//! LLVM lowering implementations for HUGR prelude operations.

use std::hash::{DefaultHasher, Hash as _, Hasher as _};
use tket::hugr::{self, llvm::inkwell};

use anyhow::{anyhow, ensure, Result};
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::extension::prelude::PreludeCodegen;
use hugr::llvm::types::TypingSession;
use inkwell::attributes::{Attribute, AttributeLoc};
use inkwell::module::Linkage;
use inkwell::types::{BasicType, IntType};
use inkwell::values::{BasicValue, BasicValueEnum, FunctionValue, StructValue};
use inkwell::AddressSpace;
use tket::hugr::{HugrView, Node};

#[derive(Clone)]
/// Codegen extension for the QIS prelude.
pub struct QISPreludeCodegen;

impl PreludeCodegen for QISPreludeCodegen {
    fn usize_type<'c>(&self, session: &TypingSession<'c, '_>) -> IntType<'c> {
        session.iw_context().i64_type()
    }

    /// Return the llvm type of [`tket::hugr::extension::prelude::qb_t`].
    fn qubit_type<'c>(&self, session: &TypingSession<'c, '_>) -> impl BasicType<'c> {
        session.iw_context().i64_type()
    }

    /// A panic ends all shots, this is done by using a return code > 1000.
    /// The lowering of the panic HUGR op shifts the return code held by the op by 1000.
    /// So valid values for `signal` in the panic HUGR op are 1-(2^31 - 1001).
    fn emit_panic<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_panic: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-(2^31 - 1001), otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let return_code = {
            let shift = ctx.iw_context().i32_type().const_int(1000, false);
            ctx.builder()
                .build_int_add(return_code.into_int_value(), shift, "shift_code")?
        };

        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    /// An exit ends the current shot using a return code `rc`, where `1 <= rc <= 1000`.
    fn emit_exit<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_exit: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-1000, otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    fn emit_print<H: HugrView<Node = Node>>(
        &self,
        _ctx: &mut EmitFuncContext<H>,
        _text: inkwell::values::BasicValueEnum,
    ) -> Result<()> {
        Ok(())
    }

    fn emit_const_error<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        err: &hugr::extension::prelude::ConstError,
    ) -> Result<BasicValueEnum<'c>> {
        let err_ty = ctx
            .llvm_type(&hugr::extension::prelude::error_type())?
            .into_struct_type();
        let signal = err_ty
            .get_field_type_at_index(0)
            .unwrap()
            .into_int_type()
            .const_int(u64::from(err.signal), false);

        let message = emit_global_string(ctx, &err.message, "e_", "EXIT:INT:")?;
        let err = err_ty.const_named_struct(&[signal.into(), message]);
        Ok(err.into())
    }

    fn emit_const_string<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        str: &hugr::extension::prelude::ConstString,
    ) -> Result<BasicValueEnum<'c>> {
        emit_global_string(ctx, str.value(), "s_", "")
    }
}

impl QISPreludeCodegen {
    fn get_panic<'c, H: HugrView<Node = Node>>(
        ctx: &EmitFuncContext<'c, '_, H>,
    ) -> Result<FunctionValue<'c>> {
        let iwc = ctx.iw_context();
        let func_type = iwc.void_type().fn_type(
            &[
                iwc.i32_type().into(),
                iwc.i8_type().ptr_type(AddressSpace::default()).into(),
            ],
            false,
        );
        let f = ctx.get_extern_func("panic", func_type)?;
        let noreturn = Attribute::get_named_enum_kind_id("noreturn");
        debug_assert!(noreturn != 0);
        f.add_attribute(
            AttributeLoc::Function,
            iwc.create_enum_attribute(noreturn, 0),
        );

        Ok(f)
    }
}

/// Emit a global string constant with a unique name.
///
/// The symbol name is constructed with `symbol_prefix` and ``str`.
///
/// The string is prefixed with `str_prefix`.
pub fn emit_global_string<'c, H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'c, '_, H>,
    str: impl AsRef<str>,
    symbol_prefix: impl AsRef<str>,
    str_prefix: impl AsRef<str>,
) -> Result<BasicValueEnum<'c>> {
    let (str, symbol_prefix, str_prefix) =
        (str.as_ref(), symbol_prefix.as_ref(), str_prefix.as_ref());
    let tagged_str = format!("{str_prefix}{str}");
    let tagged_str_bytes = tagged_str.as_bytes();
    let tagged_str_len = tagged_str_bytes.len();
    ensure!(
        tagged_str_len < 256,
        "Constant string too long: {tagged_str_len} >= 256"
    );

    let tgt_str_bytes = [&[u8::try_from(tagged_str_len)?], tagged_str_bytes].concat();
    let tgt_str_value = ctx.iw_context().const_string(&tgt_str_bytes, false);

    let global = {
        // we need a unique name for the global. We use a prefix(of length
        // 10) of the string, followed by the hash of the whole string.
        //
        // This should be enough, but for absolute guarantee of success we
        // append ".i", i an integer and find an available name.
        let hash = {
            let mut s = DefaultHasher::default();
            tgt_str_bytes.hash(&mut s);
            let h = s.finish();
            let lower32 = h as u32;
            format!("{lower32:08X}")
        };
        // Use unicode-aware character slicing to prevent slicing through multi-byte characters
        let prefix_chars: String = str.chars().take(10).collect();
        let name_prefix = format!("{symbol_prefix}{prefix_chars}.{hash}");

        (0..)
            .find_map(|i| {
                let name = format!("{name_prefix}.{i}");
                if let Some(global) = ctx.get_current_module().get_global(&name) {
                    if let Some(v) = global.get_initializer() {
                        if v == tgt_str_value {
                            // We've found a global with the same value, use that.
                            return Some(global);
                        }
                    }
                } else {
                    let global =
                        ctx.get_current_module()
                            .add_global(tgt_str_value.get_type(), None, &name);
                    global.set_initializer(&tgt_str_value);
                    global.set_linkage(Linkage::Private);
                    global.set_constant(true);
                    return Some(global);
                }
                None
            })
            .unwrap()
    };
    Ok(ctx
        .builder()
        .build_pointer_cast(
            global.as_pointer_value(),
            ctx.iw_context().i8_type().ptr_type(AddressSpace::default()),
            "",
        )?
        .as_basic_value_enum())
}

#[cfg(test)]
mod test {
    use hugr::builder::{Dataflow, DataflowHugr};
    use hugr::extension::prelude::{self, qb_t, ConstError, EXIT_OP_ID, PANIC_OP_ID};
    use hugr::extension::PRELUDE;
    use hugr::llvm::check_emission;
    use hugr::llvm::emit::test::SimpleHugrConfig;

    use hugr::llvm::test::llvm_ctx;
    use hugr::types::TypeArg;
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_panic_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        // Create a hugr that has a panic message
        let error_val = ConstError::new(42, "PANIC");
        let type_arg_q = TypeArg::Runtime(qb_t());
        let type_arg_2q = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let panic_op = PRELUDE
            .instantiate_extension_op(&PANIC_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(panic_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    fn test_exit_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        let error_val = ConstError::new(42, "EXIT");
        let type_arg_q: TypeArg = TypeArg::Runtime(qb_t());
        let type_arg_2q: TypeArg = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let exit_op = PRELUDE
            .instantiate_extension_op(&EXIT_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(exit_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }
}
