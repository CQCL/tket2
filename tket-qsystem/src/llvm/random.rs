//! LLVM lowering implementations for the "tket.qsystem.random" extension.

use tket::hugr::{self, llvm::inkwell};

use crate::extension::random::{self, RandomOp, CONTEXT_TYPE_NAME};
use anyhow::{anyhow, Result};
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::emit::EmitOpArgs;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{BasicTypeEnum, FloatType, IntType};
use inkwell::values::{BasicValueEnum, FunctionValue};
use tket::hugr::ops::ExtensionOp;
use tket::hugr::{HugrView, Node};

/// Codegen extension for `tket.qsystem.random` extension.
pub struct RandomCodegenExtension;

impl CodegenExtension for RandomCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: tket::hugr::llvm::CodegenExtsBuilder<'a, H>,
    ) -> tket::hugr::llvm::CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type(
                (random::EXTENSION_ID, CONTEXT_TYPE_NAME.to_owned()),
                |session, _| Ok(rng_context_type(session.iw_context())),
            )
            .simple_extension_op(|context, args, op| RandomEmitter(context).emit(args, op))
    }
}

fn rng_context_type(context: &Context) -> BasicTypeEnum<'_> {
    context.struct_type(&[], false).into()
}

/// Lower the `tket.qsystem.random` extension.
struct RandomEmitter<'c, 'd, 'e, H: HugrView<Node = Node>>(&'d mut EmitFuncContext<'c, 'e, H>);

impl<'c, H: HugrView<Node = Node>> RandomEmitter<'c, '_, '_, H> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn i32_type(&self) -> IntType<'c> {
        self.iw_context().i32_type()
    }

    fn i64_type(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn float_type(&self) -> FloatType<'c> {
        self.iw_context().f64_type()
    }

    fn bool_type(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn rng_context(&self) -> BasicValueEnum<'c> {
        self.iw_context().const_struct(&[], false).into()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    /// Helper function to `emit` an RNG operation.
    fn emit_op(
        &self,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        name: &str,
        func: Result<FunctionValue<'c>>,
        input_indices: &[usize],
    ) -> Result<()> {
        let inputs: Vec<_> = input_indices.iter().map(|&i| args.inputs[i]).collect();
        let result = self
            .builder()
            .build_call(
                func?,
                &inputs.iter().map(|&v| v.into()).collect::<Vec<_>>(),
                name,
            )?
            .try_as_basic_value()
            .unwrap_left();
        args.outputs
            .finish(self.builder(), [result, self.rng_context()])
    }

    /// Function to help lower the `tket.qsystem.random` extension.
    fn emit(&self, args: EmitOpArgs<'c, '_, ExtensionOp, H>, op: RandomOp) -> Result<()> {
        match op {
            RandomOp::RandomInt => {
                let fn_random_int = self
                    .0
                    .get_extern_func("random_int", self.i32_type().fn_type(&[], false));
                self.emit_op(args, "rint", fn_random_int, &[])
            }
            RandomOp::RandomFloat => {
                let fn_random_float = self
                    .0
                    .get_extern_func("random_float", self.float_type().fn_type(&[], false));
                self.emit_op(args, "rfloat", fn_random_float, &[])
            }
            RandomOp::RandomIntBounded => {
                let fn_random_int_bounded = self.0.get_extern_func(
                    "random_rng",
                    self.i32_type().fn_type(&[self.i32_type().into()], false),
                );
                self.emit_op(args, "rintb", fn_random_int_bounded, &[1])
            }
            RandomOp::NewRNGContext => {
                let fn_random_seed = self.0.get_extern_func(
                    "random_seed",
                    self.iw_context()
                        .void_type()
                        .fn_type(&[self.i64_type().into()], false),
                )?;
                let [seed] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("NewRNGContext expects a seed argument"))?;
                self.builder()
                    .build_call(fn_random_seed, &[seed.into()], "rseed")?;
                args.outputs.finish(
                    self.builder(),
                    [self.bool_type().const_int(1, false).into()],
                )
            }
            RandomOp::DeleteRNGContext => args.outputs.finish(self.builder(), []),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::extension::random::RandomOp;
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::random_int(1, RandomOp::RandomInt)]
    #[case::random_float(2, RandomOp::RandomFloat)]
    #[case::random_int_bounded(3, RandomOp::RandomIntBounded)]
    #[case::new_rng_context(4, RandomOp::NewRNGContext)]
    #[case::delete_rng_context(5, RandomOp::DeleteRNGContext)]
    fn emit_random_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: RandomOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(RandomCodegenExtension)
                .add_default_int_extensions()
                .add_float_extensions()
                .add_logic_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
