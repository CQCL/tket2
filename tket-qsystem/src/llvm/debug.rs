//! LLVM lowering implementations for "tket.debug" extension.

use crate::llvm::prelude::emit_global_string;
use anyhow::{anyhow, bail, Result};
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::{EmitFuncContext, EmitOpArgs};
use hugr::llvm::inkwell::AddressSpace;
use hugr::llvm::CodegenExtsBuilder;
use hugr::ops::ExtensionOp;
use hugr::{HugrView, Node};
use tket::extension::debug::{StateResult, DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID};

use super::array_utils::{struct_1d_arr_alloc, struct_1d_arr_ptr_t, ArrayLowering, ElemType};

static TAG_PREFIX: &str = "USER:";

/// Codegen extension for debug functionality.
#[derive(Default)]
pub struct DebugCodegenExtension<AL: ArrayLowering> {
    array_lowering: AL,
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Creates a new [DebugCodegenExtension] with specified array lowering.
    pub const fn new(array_lowering: AL) -> Self {
        Self { array_lowering }
    }
}

impl<AL: ArrayLowering> CodegenExtension for DebugCodegenExtension<AL> {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder.extension_op(DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID, {
            move |context, args| self.emit_state_result(context, args)
        })
    }
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Lower the `debug` `StateResult` op.
    fn emit_state_result<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    ) -> Result<()> {
        let builder = ctx.builder();

        // Types (qubits are just i64s).
        let iw_ctx = ctx.iw_context();
        let void_t = iw_ctx.void_type();
        let i8_ptr_t = iw_ctx.i8_type().ptr_type(AddressSpace::default());
        let i64_t = iw_ctx.i64_type();

        // Tag arguments.
        let state_result = StateResult::from_extension_op(args.node().as_ref())?;
        let tag = state_result.tag;
        if tag.is_empty() {
            bail!("Empty state result tag received");
        }
        let tag_ptr = emit_global_string(ctx, tag, "res_", format!("{TAG_PREFIX}STATE:"))?;
        let tag_len = {
            let mut l = builder
                .build_load(tag_ptr.into_pointer_value(), "tag_len")?
                .into_int_value();
            if l.get_type() != i64_t {
                l = builder.build_int_z_extend(l, i64_t, "tag_len")?;
            }
            l
        };

        // Qubit array argument.
        let array_len = state_result.num_qubits;
        let [qubits] = args
            .inputs
            .try_into()
            .map_err(|_| anyhow!(format!("StateResult expects a qubit array argument")))?;
        let qubits_array = self.array_lowering.array_to_ptr(builder, qubits)?;
        let (qubits_ptr, _) = struct_1d_arr_alloc(
            iw_ctx,
            builder,
            array_len.try_into()?,
            &ElemType::Int,
            qubits_array,
        )?;

        // Build the function call.
        let fn_state_result = ctx.get_extern_func(
            "print_state_result",
            void_t.fn_type(
                &[
                    i8_ptr_t.into(),
                    i64_t.into(),
                    struct_1d_arr_ptr_t(iw_ctx, &ElemType::Int).into(),
                ],
                false,
            ),
        )?;

        builder.build_call(
            fn_state_result,
            &[tag_ptr.into(), tag_len.into(), qubits_ptr.into()],
            "print_state_result",
        )?;
        args.outputs.finish(builder, [qubits])
    }
}

#[cfg(test)]
#[allow(deprecated)] // TODO remove after switching to new array lowering
mod test {
    use tket::extension::debug::StateResult;

    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;

    use crate::llvm::array_utils::DEFAULT_HEAP_ARRAY_LOWERING;
    use crate::llvm::array_utils::DEFAULT_STACK_ARRAY_LOWERING;
    use crate::llvm::prelude::QISPreludeCodegen;

    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::state_result(1, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::state_result(2, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_HEAP_ARRAY_LOWERING)]
    fn emit_debug_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: StateResult,
        #[case] array_lowering: &'static (impl ArrayLowering + Clone),
    ) {
        let pcg = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| {
            ceb.add_extension(DebugCodegenExtension::new(array_lowering.clone()))
                .add_extension(array_lowering.codegen_extension())
                .add_prelude_extensions(pcg.clone())
                .add_default_int_extensions()
                .add_float_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
