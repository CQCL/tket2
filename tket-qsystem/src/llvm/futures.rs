//! LLVM lowering implementations for "tket.futures" operations.

use crate::extension::futures::{self, FutureOp, FutureOpDef, FUTURE_TYPE_NAME};
use anyhow::{anyhow, Result};
use hugr::extension::prelude::bool_t;
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::ops::{ExtensionOp, Value};
use hugr::std_extensions::arithmetic::int_types::INT_TYPES;
use hugr::types::TypeArg;
use hugr::{HugrView, Node};
use hugr_llvm::custom::CodegenExtension;
use hugr_llvm::emit::func::EmitFuncContext;
use hugr_llvm::emit::{emit_value, EmitOpArgs};
use hugr_llvm::CodegenExtsBuilder;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{BasicTypeEnum, IntType};
use inkwell::values::FunctionValue;
use tket::hugr::{
    self,
    llvm::{self as hugr_llvm, inkwell},
};

/// Codegen extension for futures
pub struct FuturesCodegenExtension;

impl CodegenExtension for FuturesCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type(
                (futures::EXTENSION_ID, FUTURE_TYPE_NAME.to_owned()),
                |session, hugr_type| {
                    match (hugr_type.name().as_str(), hugr_type.args()) {
                        // For now, we only support future bools
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == bool_t() => {
                            Ok(future_type(session.iw_context()))
                        }
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == INT_TYPES[6] => {
                            Ok(future_type(session.iw_context()))
                        }
                        _ => Err(anyhow!(
                            "FuturesCodegenExtension: Unsupported type: {}",
                            hugr_type
                        )),
                    }
                },
            )
            .simple_extension_op::<FutureOpDef>(|context, args, _op| {
                FuturesEmitter(context).emit(args)
            })
    }
}

/// The LLVM type representing a future handle.
pub fn future_type(context: &Context) -> BasicTypeEnum<'_> {
    // The runtime represent a future handle as an i64
    context.i64_type().into()
}

struct FuturesEmitter<'c, 'd, 'e, H: HugrView<Node = Node>>(&'d mut EmitFuncContext<'c, 'e, H>);

impl<'c, H: HugrView<Node = Node>> FuturesEmitter<'c, '_, '_, H> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn ll_future_type(&self) -> BasicTypeEnum<'c> {
        future_type(self.iw_context())
    }

    fn ll_bool_type(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn ll_uint_type(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    fn get_func_inc_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___inc_future_refcount", func_type)
    }

    fn get_func_dec_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___dec_future_refcount", func_type)
    }

    fn get_func_read_bool(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_bool_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_bool", func_type)
    }

    fn get_func_read_uint(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_uint_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_uint", func_type)
    }

    fn emit(&mut self, args: EmitOpArgs<'c, '_, ExtensionOp, H>) -> Result<()> {
        let future_op = FutureOp::from_optype(&args.node().generalise()).unwrap();
        let op = &future_op.op;
        let typ = &future_op.typ;
        match (op, typ) {
            (FutureOpDef::Dup, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Dup expects a single input"))?;
                let func = self.get_func_inc_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "inc_refcount")?;
                args.outputs.finish(self.builder(), [arg, arg])
            }
            (FutureOpDef::Free, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Free expects a single input"))?;
                let func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [])
            }
            (FutureOpDef::Read, ty) if *ty == bool_t() => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_bool()?;
                let result_i1 = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_bool")?
                    .try_as_basic_value()
                    .unwrap_left()
                    .into_int_value();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                let true_val = emit_value(self.0, &Value::true_val())?;
                let false_val = emit_value(self.0, &Value::false_val())?;
                let result = self
                    .builder()
                    .build_select(result_i1, true_val, false_val, "measure")?;
                args.outputs.finish(self.builder(), [result])
            }
            (FutureOpDef::Read, ty) if *ty == INT_TYPES[6] => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_uint()?;
                let result = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_uint")?
                    .try_as_basic_value()
                    .unwrap_left();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [result])
            }
            _ => Err(anyhow!(
                "Unsupported future operation: {:?} with type: {}",
                op,
                typ
            )),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::std_extensions::arithmetic::int_types::int_type;
    use hugr_llvm::check_emission;
    use hugr_llvm::test::llvm_ctx;
    use hugr_llvm::test::single_op_hugr;
    use hugr_llvm::test::TestContext;

    use super::*;
    #[rstest::rstest]
    #[case::read_bool(1,FutureOp { op: FutureOpDef::Read, typ: bool_t() })]
    #[case::dup_bool(2,FutureOp { op: FutureOpDef::Dup, typ: bool_t() })]
    #[case::free_bool(3,FutureOp { op: FutureOpDef::Free, typ: bool_t() })]
    #[case::read_int(4,FutureOp { op: FutureOpDef::Read, typ: int_type(6) })]
    #[case::dup_int(5,FutureOp { op: FutureOpDef::Dup, typ: int_type(6) })]
    #[case::free_int(6,FutureOp { op: FutureOpDef::Free, typ: int_type(6) })]
    fn emit_futures_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: FutureOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(FuturesCodegenExtension)
                .add_default_int_extensions()
        });
        let hugr = single_op_hugr(op.to_extension_op().unwrap().into());
        check_emission!(hugr, llvm_ctx);
    }
}
