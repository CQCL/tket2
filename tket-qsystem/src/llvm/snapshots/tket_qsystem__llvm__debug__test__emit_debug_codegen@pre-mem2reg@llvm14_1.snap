---
source: tket-qsystem/src/llvm/debug.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

@res_test_state.900F7606.0 = private constant [29 x i8] c"\1CUSER:STATE:test_state_result"

define private [2 x i64] @_hl.main.1([2 x i64] %0) {
alloca_block:
  %"0" = alloca [2 x i64], align 8
  %"2_0" = alloca [2 x i64], align 8
  %"4_0" = alloca [2 x i64], align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store [2 x i64] %0, [2 x i64]* %"2_0", align 4
  %"2_01" = load [2 x i64], [2 x i64]* %"2_0", align 4
  %tag_len = load i8, i8* getelementptr inbounds ([29 x i8], [29 x i8]* @res_test_state.900F7606.0, i32 0, i32 0), align 1
  %tag_len2 = zext i8 %tag_len to i64
  %1 = alloca i64, i32 2, align 8
  %2 = bitcast i64* %1 to [2 x i64]*
  store [2 x i64] %"2_01", [2 x i64]* %2, align 4
  %out_arr_alloca = alloca <{ i32, i32, i64*, i1* }>, align 8
  %x_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 0
  %y_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 1
  %arr_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 2
  %mask_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 3
  %3 = alloca i1, i32 2, align 1
  %4 = bitcast i1* %3 to [2 x i1]*
  store [2 x i1] zeroinitializer, [2 x i1]* %4, align 1
  store i32 2, i32* %x_ptr, align 4
  store i32 1, i32* %y_ptr, align 4
  store i64* %1, i64** %arr_ptr, align 8
  store i1* %3, i1** %mask_ptr, align 8
  %5 = load <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, align 1
  call void @print_state_result(i8* getelementptr inbounds ([29 x i8], [29 x i8]* @res_test_state.900F7606.0, i32 0, i32 0), i64 %tag_len2, <{ i32, i32, i64*, i1* }>* %out_arr_alloca)
  store [2 x i64] %"2_01", [2 x i64]* %"4_0", align 4
  %"4_03" = load [2 x i64], [2 x i64]* %"4_0", align 4
  store [2 x i64] %"4_03", [2 x i64]* %"0", align 4
  %"04" = load [2 x i64], [2 x i64]* %"0", align 4
  ret [2 x i64] %"04"
}

declare void @print_state_result(i8*, i64, <{ i32, i32, i64*, i1* }>*)
